### PR TITLE
ingester: recover segments on startup

### LINKED ISSUE #1: ingest: file log should recover .active segments on startup
Otherwise, if the ingesters crash, records that haven't yet been flushed are kind of in limbo. This should be a simple fix.

### PR DESCRIPTION
**Pull request by peterbourgon**
If an ingester crashes, it may leave .active and .pending segment files in an abandoned state. On startup, it can look for such abandoned segments and recover them, by making them .flushed again. This may confuse stores, but that confusion will only result in canceled transactions and duplicated effort—no problem.

The recovery process will just be renaming .active and .pending segments to .flushed. If that happens on an active data dir, then bad things will happen! So, as a first step, we'll implement a data dir locking mechanism, with the tried-and-true LOCK file pattern. Technically we just need it for the ingesters, but since it generalizes pretty well, I'll also add the same locking behavior for stores in this PR, too.

Closes #1.

### CODE REVIEW

File: pkg/fs/lock.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	"github.com/pkg/errors"
+)
+
+const lockFilename = "LOCK"
+
+// ClaimLock on the specified path.
+// Successful claims must eventually be released.
+func ClaimLock(fs Filesystem, path string) error {
+	lockFile := filepath.Join(path, lockFilename)
+	if fs.Exists(lockFile) {
```
Comment (peterbourgon) [inline pkg/fs/lock.go]: Yep, excellent observation. Unfortunately, realFilesystem.Create calls os.Create, which "creates the named file with mode 0666 (before umask), **truncating it if it already exists.**". And I didn't immediately see a non-racy way to get Create-if-not-exists semantics. Perhaps I'll have to upgrade to something like [prometheus/util/flock](https://godoc.org/github.com/prometheus/prometheus/util/flock)...

Comment (danp) [inline pkg/fs/lock.go]: Oh yes, that would make me feel much better. :) Was going to say too that this kind of locking is racy in general.

Comment (peterbourgon) [inline pkg/fs/lock.go]: Computers :(

File: pkg/fs/real.go [hunk 2]
```diff
@@ -78,6 +80,24 @@ func (realFilesystem) Walk(root string, walkFn filepath.WalkFunc) error {
...
+type deletingReleaser struct {
+	path string
+	r    Releaser
+}
+
+func (dr deletingReleaser) Release() error {
+	if err := dr.r.Release(); err != nil {
+		return err
+	}
+	return os.Remove(dr.path)
```
Comment (danp) [inline pkg/fs/real.go]: Maybe I don't have things straight but this still feels racy. Should the flock package remove this while holding the lock, like around [here](https://github.com/prometheus/prometheus/blob/7eb849e6a82094b649087e36d2ce5132fc798ef6/util/flock/flock_unix.go#L31)? Surprised there's no *os.File.Remove to reference here.

Comment (danp) [inline pkg/fs/real.go]: Or maybe this Remove should be moved up above the call to Release?

Comment (peterbourgon) [inline pkg/fs/real.go]: Remove before Release works for me!

**Comment by danp** [comment]
LGTM

**Comment by danp** [comment]
Still LGTM!

**Comment by peterbourgon** [comment]
😎  Thanks for the reviews, @danp!

**Pull request merged**